File: src/web/Backend/src/routes/explorer_routes.js
const fs = require("node:fs");
const fsPromises = require("node:fs/promises");
const path = require("node:path");
const router = require("express").Router();
const { ExplorerService } = require("../services/explorerService");
const { getConfig } = require("../config/env");
const { createRateLimiter } = require("../utils/rateLimit");
const { guessKind, isHex64 } = require("../utils/searchKind");

const isArtId = (s) => /^graf[0-9a-fA-F]{60}$/.test(s || "");

const cfg = getConfig();
const svc = new ExplorerService({ nodeHost: cfg.nodeHost, nodePort: cfg.nodePort });
const projectRoot = path.resolve(__dirname, "..", "..", "..", "..", "..");
const cacheDir = path.resolve(projectRoot, "data", "web", "graffiti_cache");
const GRAFFITI_CACHE_TTL_MS = 6 * 60 * 60 * 1000;
const GRAFFITI_CACHE_SWEEP_MS = 15 * 60 * 1000;
let lastGraffitiCacheSweep = 0;

const searchLimiter = createRateLimiter({ windowMs: 60 * 1000, max: 20 });
const graffitiMediaLimiter = createRateLimiter({ windowMs: 60 * 1000, max: 40 });

const touchFile = async (filePath) => {
  try {
    const now = new Date();
    await fsPromises.utimes(filePath, now, now);
  } catch (err) {
    console.warn("Failed to touch file:", err);
  }
};

const cleanupGraffitiCache = async () => {
  const now = Date.now();
  if (now - lastGraffitiCacheSweep < GRAFFITI_CACHE_SWEEP_MS) return;
  lastGraffitiCacheSweep = now;
  try {
    await fsPromises.access(cacheDir);
  } catch {
    return;
  }
  let entries;
  try {
    entries = await fsPromises.readdir(cacheDir);
  } catch (err) {
    console.warn("Failed to read cache directory:", err);
    return;
  }
  for (const entry of entries) {
    const fullPath = path.join(cacheDir, entry);
    let stat;
    try {
      stat = await fsPromises.stat(fullPath);
    } catch (err) {
      console.warn("Failed to stat file:", err);
      continue;
    }
    if (!stat.isFile()) continue;
    if (now - stat.mtimeMs > GRAFFITI_CACHE_TTL_MS) {
      try {
        await fsPromises.unlink(fullPath);
      } catch (err) {
        console.warn("Failed to unlink file:", err);
      }
    }
  }
};

const resolveCachePath = (cachePath) => {
  if (!cachePath) return null;
  if (path.isAbsolute(cachePath)) return cachePath;
  return path.resolve(projectRoot, cachePath);
};

const inferMediaType = (meta, filePath) => {
  const mime = meta?.mime || meta?.mime_type;
  if (mime) {
    if (String(mime).includes("video")) return "video/mp4";
    if (String(mime).includes("image")) return "image/jpeg";
  }
  const ext = path.extname(filePath || "").toLowerCase();
  if (ext === ".mp4") return "video/mp4";
  if (ext === ".jpg" || ext === ".jpeg") return "image/jpeg";
  return "application/octet-stream";
};

const findCachedFile = async (artId) => {
  try {
    await fsPromises.access(cacheDir);
  } catch {
    return null;
  }
  for (const ext of [".jpg", ".jpeg", ".mp4", ".bin"]) {
    const candidate = path.join(cacheDir, `${artId}${ext}`);
    try {
      await fsPromises.access(candidate);
      return candidate;
    } catch {
      // file not found, continue
    }
  }
  return null;
};

const parseRangeHeader = (rangeHeader, size) => {
  if (!rangeHeader) return null;
  const match = /^bytes=(\d*)-(\d*)$/.exec(rangeHeader);
  if (!match) {
    return { invalid: true };
  }
  const start = match[1] ? Number(match[1]) : 0;
  let end = match[2] ? Number(match[2]) : size - 1;
  if (Number.isNaN(start) || Number.isNaN(end) || start > end || start >= size) {
    return { invalid: true };
  }
  end = Math.min(end, size - 1);
  return { start, end };
};

router.get("/receipt", async (req, res, next) => {
  try {
    const txid = req.query.txid;
    if (!txid) {
      return res.status(400).json({ error: "missing_txid" });
    }
    if (!isHex64(txid)) {
      return res.status(400).json({ error: "invalid_txid" });
    }

    const data = await svc.getReceipt(txid);
    if (data.status === "error") {
      return res.status(400).json({ error: data.message });
    }
    
    res.json({ status: "ok", data });
  } catch (err) {
    next(err);
  }
});

router.get("/history_book", async (req, res, next) => {
  try {
    const address = req.query.address;
    if (!address) {
      return res.status(400).json({ error: "missing_address" });
    }
    if (!/^tsar[0-9a-zA-Z]{16,}$/.test(address)) {
      return res.status(400).json({ error: "invalid_address" });
    }

    const data = await svc.getHistoryBook(address);
    if (data.status === "error") {
      return res.status(400).json({ error: data.message });
    }
    
    res.json({ status: "ok", data });
  } catch (err) {
    next(err);
  }
});

router.get("/network", async (_req, res, next) => {
  try {
    const snap = await svc.getNetwork();
    res.json({ status: "ok", data: snap });
  } catch (err) {
    next(err);
  }
});

router.get("/blocks", async (req, res, next) => {
  try {
    const limit = Math.min(Math.max(Number(req.query.limit) || 10, 1), 500);
    const startRaw = req.query.start ?? req.query.start_height ?? req.query.height;
    const startHeight = 
      startRaw === undefined || startRaw === null || startRaw === ""
        ? null
        : Number(startRaw);
    
    // Parameter untuk prefer database
    const preferDatabase = req.query.prefer_database === 'true';
    
    const data = await svc.getBlockRange({ 
      startHeight, 
      limit,
      source: preferDatabase ? 'database' : 'auto'
    });
    
    res.json({ status: "ok", data });
  } catch (err) {
    next(err);
  }
});

router.get("/block/:id", async (req, res, next) => {
  try {
    const id = req.params.id;
    const isHeight = /^\d{1,7}$/.test(id);
    if (!isHeight && !isHex64(id)) {
      return res.status(400).json({ error: "invalid_block_id" });
    }
    const data = await svc.getBlock(id);
    res.json({ status: "ok", data });
  } catch (err) {
    next(err);
  }
});

router.get("/tx/:id", async (req, res, next) => {
  try {
    const id = req.params.id;
    if (!isHex64(id)) {
      return res.status(400).json({ error: "invalid_txid" });
    }
    const data = await svc.getTx(id);
    res.json({ status: "ok", data });
  } catch (err) {
    next(err);
  }
});

router.get("/address/:addr", async (req, res, next) => {
  try {
    const addr = req.params.addr;
    if (!/^tsar[0-9a-zA-Z]{16,}$/.test(addr)) {
      return res.status(400).json({ error: "invalid_address" });
    }
    const data = await svc.getAddress(addr);
    res.json({ status: "ok", data });
  } catch (err) {
    next(err);
  }
});

router.get("/graffiti/:artId", async (req, res, next) => {
  try {
    const artId = req.params.artId;
    if (!isArtId(artId)) {
      return res.status(400).json({ error: "invalid_art_id" });
    }
    const data = await svc.getGraffiti(artId);
    if (!data) {
      return res.status(404).json({ error: "not_found" });
    }
    res.json({ status: "ok", data });
  } catch (err) {
    next(err);
  }
});

const STREAM_THRESHOLD_BYTES = 10 * 1024 * 1024;
const STREAM_CHUNK_BYTES = 4 * 1024 * 1024;

const serveLocalFile = async (req, res, next, filePath, meta = null) => {
  try {
    await fsPromises.access(filePath);
    touchFile(filePath).catch((err) => console.warn("Touch file error:", err));
    const stat = await fsPromises.stat(filePath);
    const size = stat.size;
    const type = inferMediaType(meta, filePath);
    res.setHeader("Content-Type", type);
    res.setHeader("Cache-Control", "public, max-age=300");
    res.setHeader("Accept-Ranges", "bytes");

    const range = parseRangeHeader(req.headers.range, size);
    if (range) {
      if (range.invalid) {
        res.status(416);
        res.setHeader("Content-Range", `bytes */${size}`);
        res.end();
        return true;
      }
      res.status(206);
      res.setHeader("Content-Range", `bytes ${range.start}-${range.end}/${size}`);
      res.setHeader("Content-Length", String(range.end - range.start + 1));
      const stream = fs.createReadStream(filePath, { start: range.start, end: range.end });
      stream.on("error", next);
      stream.pipe(res);
      return true;
    }

    res.setHeader("Content-Length", String(size));
    const stream = fs.createReadStream(filePath);
    stream.on("error", next);
    stream.pipe(res);
    return true;
  } catch (err) {
    console.warn("Failed to serve local cached file:", err);
    return false;
  }
};

const tryServeFromServiceCache = async (artId, req, res, next) => {
  const info = await svc.getGraffitiMediaInfo(artId);
  if (info?.status === "ok" && info?.cache_path) {
    const resolvedPath = resolveCachePath(info.cache_path);
    if (resolvedPath) {
      return await serveLocalFile(req, res, next, resolvedPath, info.meta);
    }
  }
  return false;
};

const streamGraffitiChunks = async (req, res, artId, totalSize, meta) => {
  const type = inferMediaType(meta, meta?.filename || artId);
  res.setHeader("Content-Type", type);
  res.setHeader("Cache-Control", "public, max-age=300");
  res.setHeader("Accept-Ranges", "bytes");

  let start = 0;
  let end = totalSize > 0 ? totalSize - 1 : 0;

  const range = totalSize > 0 ? parseRangeHeader(req.headers.range, totalSize) : null;
  if (range) {
    if (range.invalid) {
      res.status(416);
      res.setHeader("Content-Range", `bytes */${totalSize}`);
      return res.end();
    }
    start = range.start;
    end = range.end;
    res.status(206);
    res.setHeader("Content-Range", `bytes ${start}-${end}/${totalSize}`);
    res.setHeader("Content-Length", String(end - start + 1));
  } else if (totalSize > 0) {
    res.setHeader("Content-Length", String(totalSize));
  }

  let clientAborted = false;
  req.on("close", () => {
    clientAborted = true;
  });

  let currOffset = start;
  const targetEnd = totalSize > 0 ? end : Number.MAX_SAFE_INTEGER;

  while (currOffset <= targetEnd && !clientAborted) {
    const want = totalSize > 0
      ? Math.min(STREAM_CHUNK_BYTES, targetEnd - currOffset + 1)
      : STREAM_CHUNK_BYTES;

    const chunkResp = await svc.getGraffitiChunk(artId, currOffset, want);
    if (chunkResp?.status !== "ok" || !chunkResp?.data_b64) {
      break;
    }

    const buf = Buffer.from(chunkResp.data_b64, "base64");
    if (buf.length === 0) break;

    const canContinue = res.write(buf);
    currOffset += buf.length;

    if (!canContinue && !clientAborted) {
      await new Promise((resolve) => res.once("drain", resolve));
    }

    if (chunkResp.eof) break;
  }

  res.end();
};

router.get("/graffiti/:artId/media", graffitiMediaLimiter, async (req, res, next) => {
  try {
    cleanupGraffitiCache().catch((err) => console.warn("Cache cleanup error:", err));
    const artId = req.params.artId;
    if (!isArtId(artId)) {
      return res.status(400).json({ error: "invalid_art_id" });
    }

    const filePath = await findCachedFile(artId);
    if (filePath && (await serveLocalFile(req, res, next, filePath))) {
      return;
    }

    const metaResp = await svc.getGraffitiMediaMeta(artId);
    if (metaResp?.status !== "ok") {
      return res.status(404).json({ error: "media_not_found" });
    }

    const meta = metaResp.meta || {};
    const totalSize = Number(meta.size_bytes || meta.size || metaResp.size_bytes || 0);

    // Smart Caching: file size <= 10MB -> Full download to disk cache
    if (totalSize > 0 && totalSize <= STREAM_THRESHOLD_BYTES) {
      const served = await tryServeFromServiceCache(artId, req, res, next);
      if (served) return;
    }

    // File size > 10MB -> On-demand HTTP streaming (4MB chunks)
    await streamGraffitiChunks(req, res, artId, totalSize, meta);
  } catch (err) {
    next(err);
  }
});

router.get("/graffiti", async (req, res, next) => {
  try {
    const limit = Math.min(Math.max(Number(req.query.limit) || 24, 1), 100);
    const offset = Math.max(Number(req.query.offset) || 0, 0);
    const data = await svc.getGraffitiPosts({ limit, offset });
    res.json({ status: "ok", data });
  } catch (err) {
    next(err);
  }
});

router.get("/search", searchLimiter, async (req, res, next) => {
  try {
    const query = (req.query.q || "").trim();
    if (!query) return res.status(400).json({ error: "missing_query" });
    const inferred = guessKind(query);
    const result = await svc.search(query);
    if (!result.data) {
      return res.status(404).json({ status: "not_found", kind: inferred });
    }
    res.json({ status: "ok", kind: result.kind, data: result.data });
  } catch (err) {
    next(err);
  }
});

router.post("/prefetch-blocks", async (_req, res, next) => {
  try {
    // Trigger prefetch di Python RPC client
    await rpcCall("prefetch_blocks", null, cfg.nodeHost, cfg.nodePort);
    res.json({ status: "ok", message: "Prefetch started" });
  } catch (err) {
    next(err);
  }
});

module.exports = router;

File: src/web/Backend/src/routes/explorer_routes.py
# SPDX-License-Identifier: MIT
# Copyright (c) 2025 Tsar Studio
# Part of TsarChain — see LICENSE

import os
import re
import time
import threading
from typing import Dict, Any, Optional, Tuple

from web.Backend.src.utils.search_kind import guess_kind, is_hex64
from web.Backend.src.services.explorer_service import ExplorerService
from web.Backend.src.core import main_web

_ART_ID_REGEX = re.compile(r"^graf[0-9a-fA-F]{60}$")
_ADDRESS_REGEX = re.compile(r"^tsar[0-9a-zA-Z]{16,}$")
_BLOCK_HEIGHT_REGEX = re.compile(r"^\d{1,7}$")
_RANGE_REGEX = re.compile(r"^bytes=(\d*)-(\d*)$")

_CURRENT_DIR = os.path.dirname(os.path.abspath(__file__))
_PROJECT_ROOT = os.path.abspath(os.path.join(_CURRENT_DIR, "..", "..", "..", ".."))
CACHE_DIR = os.path.abspath(os.path.join(_PROJECT_ROOT, "data", "web", "graffiti_cache"))

GRAFFITI_CACHE_TTL_SEC = 6 * 60 * 60
GRAFFITI_CACHE_SWEEP_SEC = 15 * 60
STREAM_THRESHOLD_BYTES = 10 * 1024 * 1024
STREAM_CHUNK_BYTES = 4 * 1024 * 1024

_last_graffiti_cache_sweep = 0.0
_cache_sweep_lock = threading.Lock()


class ExplorerRoutes:
    def __init__(self, service: ExplorerService, host: str = "127.0.0.1", port: int = 19000):
        self.svc = service
        self.node_host = host
        self.node_port = port


    def handle_receipt(self, query_params: Dict[str, str]) -> Tuple[int, Dict[str, Any]]:
        txid = query_params.get("txid")
        if not txid:
            return 400, {"error": "missing_txid"}
        if not is_hex64(txid):
            return 400, {"error": "invalid_txid"}

        data = self.svc.get_receipt(txid)
        if data and data.get("status") == "error":
            return 400, {"error": data.get("message", "receipt_error")}

        return 200, {"status": "ok", "data": data}


    def handle_history_book(self, query_params: Dict[str, str]) -> Tuple[int, Dict[str, Any]]:
        address = query_params.get("address")
        if not address:
            return 400, {"error": "missing_address"}
        if not _ADDRESS_REGEX.match(address):
            return 400, {"error": "invalid_address"}

        data = self.svc.get_history_book(address)
        if data and data.get("status") == "error":
            return 400, {"error": data.get("message", "history_book_error")}

        return 200, {"status": "ok", "data": data}


    def handle_network(self) -> Tuple[int, Dict[str, Any]]:
        snap = self.svc.get_network()
        return 200, {"status": "ok", "data": snap}


    def handle_blocks(self, query_params: Dict[str, str]) -> Tuple[int, Dict[str, Any]]:
        limit_raw = query_params.get("limit", "10")
        try:
            limit = min(max(int(limit_raw), 1), 500)
        except (ValueError, TypeError):
            limit = 10

        start_raw = query_params.get("start") or query_params.get("start_height") or query_params.get("height")
        start_height = None
        if start_raw is not None and str(start_raw).strip() != "":
            try:
                start_height = int(start_raw)
            except (ValueError, TypeError):
                start_height = None

        prefer_database = query_params.get("prefer_database") == "true"
        source = "database" if prefer_database else "auto"

        data = self.svc.get_block_range(start_height=start_height, limit=limit, source=source)
        return 200, {"status": "ok", "data": data}


    def handle_block(self, block_id: str) -> Tuple[int, Dict[str, Any]]:
        is_height = bool(_BLOCK_HEIGHT_REGEX.match(block_id))
        if not is_height and not is_hex64(block_id):
            return 400, {"error": "invalid_block_id"}

        data = self.svc.get_block(block_id)
        return 200, {"status": "ok", "data": data}


    def handle_tx(self, txid: str) -> Tuple[int, Dict[str, Any]]:
        if not is_hex64(txid):
            return 400, {"error": "invalid_txid"}

        data = self.svc.get_tx(txid)
        return 200, {"status": "ok", "data": data}


    def handle_address(self, addr: str) -> Tuple[int, Dict[str, Any]]:
        if not _ADDRESS_REGEX.match(addr):
            return 400, {"error": "invalid_address"}

        data = self.svc.get_address(addr)
        return 200, {"status": "ok", "data": data}


    def handle_graffiti_detail(self, art_id: str) -> Tuple[int, Dict[str, Any]]:
        if not is_art_id(art_id):
            return 400, {"error": "invalid_art_id"}

        data = self.svc.get_graffiti(art_id)
        if not data:
            return 404, {"error": "not_found"}

        return 200, {"status": "ok", "data": data}


    def handle_graffiti_list(self, query_params: Dict[str, str]) -> Tuple[int, Dict[str, Any]]:
        try:
            limit = min(max(int(query_params.get("limit", "24")), 1), 100)
        except (ValueError, TypeError):
            limit = 24
        try:
            offset = max(int(query_params.get("offset", "0")), 0)
        except (ValueError, TypeError):
            offset = 0

        data = self.svc.get_graffiti_posts(limit=limit, offset=offset)
        return 200, {"status": "ok", "data": data}


    def handle_search(self, query_params: Dict[str, str]) -> Tuple[int, Dict[str, Any]]:
        query = (query_params.get("q") or "").strip()
        if not query:
            return 400, {"error": "missing_query"}

        inferred = guess_kind(query)
        result = self.svc.search(query)
        if not result or not result.get("data"):
            return 404, {"status": "not_found", "kind": inferred}

        return 200, {"status": "ok", "kind": result.get("kind"), "data": result.get("data")}


    def handle_prefetch_blocks(self) -> Tuple[int, Dict[str, Any]]:
        main_web.dispatch_rpc("prefetch_blocks", None, self.node_host, self.node_port)
        return 200, {"status": "ok", "message": "Prefetch started"}


# =============================================================================
# =============================================================================


def is_art_id(s: str | None) -> bool:
    if not s or type(s) is not str:
        return False
    return bool(_ART_ID_REGEX.match(s))


def touch_file(file_path: str) -> None:
    try:
        now = time.time()
        os.utime(file_path, (now, now))
    except Exception:
        pass


def cleanup_graffiti_cache() -> None:
    global _last_graffiti_cache_sweep
    now = time.time()
    if now - _last_graffiti_cache_sweep < GRAFFITI_CACHE_SWEEP_SEC:
        return
    with _cache_sweep_lock:
        if now - _last_graffiti_cache_sweep < GRAFFITI_CACHE_SWEEP_SEC:
            return
        _last_graffiti_cache_sweep = now

        if not os.path.isdir(CACHE_DIR):
            return
        try:
            for entry in os.listdir(CACHE_DIR):
                full_path = os.path.join(CACHE_DIR, entry)
                if os.path.isfile(full_path):
                    try:
                        stat = os.stat(full_path)
                        if now - stat.st_mtime > GRAFFITI_CACHE_TTL_SEC:
                            os.unlink(full_path)
                    except Exception:
                        pass
        except Exception:
            pass


def resolve_cache_path(cache_path: str | None) -> Optional[str]:
    if not cache_path or type(cache_path) is not str:
        return None
    if os.path.isabs(cache_path):
        resolved = os.path.abspath(os.path.normpath(cache_path))
    else:
        resolved = os.path.abspath(os.path.join(_PROJECT_ROOT, cache_path))

    rel = os.path.relpath(resolved, _PROJECT_ROOT)
    if rel.startswith("..") or os.path.isabs(rel):
        return None
    return resolved


def infer_media_type(meta: Optional[Dict[str, Any]], file_path: str | None) -> str:
    mime = (meta.get("mime") or meta.get("mime_type")) if type(meta) is dict else None
    if mime:
        mime_str = str(mime).lower()
        if "pdf" in mime_str:
            return "application/pdf"
        if "video/mp4" in mime_str or "mp4" in mime_str:
            return "video/mp4"
        if "video/x-matroska" in mime_str or "mkv" in mime_str:
            return "video/x-matroska"
        if "image/jpeg" in mime_str or "image" in mime_str:
            return "image/jpeg"

    ext = os.path.splitext(file_path or "")[1].lower()
    if ext == ".pdf":
        return "application/pdf"
    if ext == ".mp4":
        return "video/mp4"
    if ext == ".mkv":
        return "video/x-matroska"
    if ext in (".jpg", ".jpeg"):
        return "image/jpeg"
    return "application/octet-stream"


def find_cached_file(art_id: str) -> Optional[str]:
    if not os.path.isdir(CACHE_DIR):
        return None
    for ext in [".pdf", ".jpg", ".jpeg", ".mp4", ".mkv", ".bin"]:
        candidate = os.path.join(CACHE_DIR, f"{art_id}{ext}")
        if os.path.isfile(candidate):
            return candidate
    return None


def parse_range_header(range_header: str | None, size: int) -> Optional[Dict[str, Any]]:
    if not range_header or type(range_header) is not str:
        return None
    m = _RANGE_REGEX.match(range_header.strip())
    if not m:
        return {"invalid": True}

    start_str, end_str = m.group(1), m.group(2)
    try:
        start = int(start_str) if start_str else 0
        end = int(end_str) if end_str else size - 1
    except ValueError:
        return {"invalid": True}

    if start > end or start >= size or start < 0 or end < 0:
        return {"invalid": True}

    end = min(end, size - 1)
    return {"start": start, "end": end}
